File: example_usage.py
"""
Example usage of the Task Schema Validator

This file demonstrates practical usage scenarios for the task schema validator
with constraint and success criteria extraction.
"""

from task_schema_validator import create_validator, RAGChunk


def example_1_basic_validation():
    """Example 1: Basic task validation without RAG chunks"""
    print("=" * 60)
    print("Example 1: Basic Task Validation")
    print("=" * 60)
    
    validator = create_validator()
    
    # A complete, valid task
    task = {
        'task_name': 'Implement user authentication',
        'task_type': 'implementation',
        'description': 'Add JWT-based authentication to the API',
        'parameters': {'auth_type': 'JWT', 'token_expiry': '30m'},
        'status': 'PENDING',
        'priority': 1,
        'task_notes': 'High priority security feature',
        'task_data': {
            'requirement_snippets': ['User authentication is required'],
            'requirement_type': 'functional',
            'constraints': ['Must use JWT', 'Token expiry 30 minutes'],
            'success_criteria': ['User can login', 'Token is validated'],
            'execution_complexity': 'complex',
            'project_id': 1,
            'document_ids': [1],
            'chunk_ids': [10],
            'dependencies': []
        }
    }
    
    result = validator.validate_task(task)
    
    if result.is_valid:
        print("✓ Task is valid!")
    else:
        print("✗ Task validation failed:")
        for error in result.errors:
            print(f"  - {error}")
    
    print()


def example_2_constraint_extraction():
    """Example 2: Extract constraints from RAG chunks"""
    print("=" * 60)
    print("Example 2: Constraint Extraction")
    print("=" * 60)
    
    validator = create_validator()
    
    # Task with empty constraints
    task = {
        'task_name': 'Build API endpoint',
        'task_type': 'implementation',
        'description': 'Create REST API endpoint',
        'parameters': {},
        'status': 'PENDING',
        'priority': 2,
        'task_notes': '',
        'task_data': {
            'requirement_snippets': [],
            'requirement_type': 'functional',
            'constraints': [],  # Empty - will be populated by validator
            'success_criteria': [],
            'execution_complexity': 'fast',
            'project_id': 1,
            'document_ids': [1],
            'chunk_ids': [1],
            'dependencies': []
        }
    }
    
    # RAG chunk with constraints
    chunk = RAGChunk(
        chunk_id=1,
        chunk_text="""
        The API must support HTTPS only.
        Response time must not exceed 200ms.
        The system shall comply with REST standards.
        Maximum payload size is limited to 10MB.
        """,
        document_id=1,
        project_id=1,
        has_constraints=True,
        has_success_criteria=False
    )
    
    enriched_task, result = validator.validate_and_enrich_task(task, [chunk])
    
    print(f"Extracted {len(result.extracted_constraints)} constraints:")
    for i, constraint in enumerate(result.extracted_constraints, 1):
        print(f"  {i}. {constraint}")
    
    print(f"\nTask now contains {len(enriched_task['task_data']['constraints'])} constraints")
    print()


def example_3_success_criteria_extraction():
    """Example 3: Extract success criteria from RAG chunks"""
    print("=" * 60)
    print("Example 3: Success Criteria Extraction")
    print("=" * 60)
    
    validator = create_validator()
    
    task = {
        'task_name': 'User login feature',
        'task_type': 'feature',
        'description': 'Implement user login',
        'parameters': {},
        'status': 'PENDING',
        'priority': 1,
        'task_notes': '',
        'task_data': {
            'requirement_snippets': [],
            'requirement_type': 'functional',
            'constraints': [],
            'success_criteria': [],  # Empty - will be populated
            'execution_complexity': 'complex',
            'project_id': 1,
            'document_ids': [1],
            'chunk_ids': [2],
            'dependencies': []
        }
    }
    
    # RAG chunk with success criteria
    chunk = RAGChunk(
        chunk_id=2,
        chunk_text="""
        Acceptance criteria:
        - User must be able to login with email and password
        - System should validate credentials against database
        - Invalid login must return appropriate error message
        - Successful login should generate JWT token
        Given a registered user, when they enter valid credentials,
        then they should be authenticated successfully.
        """,
        document_id=1,
        project_id=1,
        has_constraints=False,
        has_success_criteria=True
    )
    
    enriched_task, result = validator.validate_and_enrich_task(task, [chunk])
    
    print(f"Extracted {len(result.extracted_success_criteria)} success criteria:")
    for i, criterion in enumerate(result.extracted_success_criteria, 1):
        print(f"  {i}. {criterion}")
    
    print(f"\nTask now contains {len(enriched_task['task_data']['success_criteria'])} criteria")
    print()


def example_4_multiple_chunks():
    """Example 4: Aggregate data from multiple RAG chunks"""
    print("=" * 60)
    print("Example 4: Multiple RAG Chunks")
    print("=" * 60)
    
    validator = create_validator()
    
    task = {
        'task_name': 'Complete authentication system',
        'task_type': 'feature',
        'description': 'Full authentication implementation',
        'parameters': {},
        'status': 'PENDING',
        'priority': 1,
        'task_notes': 'Combines multiple requirements',
        'task_data': {
            'requirement_snippets': [],
            'requirement_type': 'functional',
            'constraints': [],
            'success_criteria': [],
            'execution_complexity': 'complex',
            'project_id': 1,
            'document_ids': [1, 2],
            'chunk_ids': [1, 2, 3],
            'dependencies': []
        }
    }
    
    chunks = [
        RAGChunk(
            chunk_id=1,
            chunk_text="System must use TLS 1.3 encryption. Passwords must be hashed with bcrypt.",
            document_id=1,
            project_id=1,
            has_constraints=True,
            has_success_criteria=False
        ),
        RAGChunk(
            chunk_id=2,
            chunk_text="Session timeout shall not exceed 30 minutes. Rate limiting is required.",
            document_id=1,
            project_id=1,
            has_constraints=True,
            has_success_criteria=False
        ),
        RAGChunk(
            chunk_id=3,
            chunk_text="User must be able to login and logout. System should validate all inputs.",
            document_id=2,
            project_id=1,
            has_constraints=False,
            has_success_criteria=True
        )
    ]
    
    enriched_task, result = validator.validate_and_enrich_task(task, chunks)
    
    print(f"Extracted from {len(chunks)} chunks:")
    print(f"  Constraints: {len(result.extracted_constraints)}")
    print(f"  Success Criteria: {len(result.extracted_success_criteria)}")
    
    print("\nConstraints:")
    for constraint in enriched_task['task_data']['constraints']:
        print(f"  - {constraint}")
    
    print("\nSuccess Criteria:")
    for criterion in enriched_task['task_data']['success_criteria']:
        print(f"  - {criterion}")
    
    print()


def example_5_validation_errors():
    """Example 5: Handling validation errors"""
    print("=" * 60)
    print("Example 5: Validation Errors")
    print("=" * 60)
    
    validator = create_validator()
    
    # Incomplete task with missing required fields
    incomplete_task = {
        'task_name': 'Incomplete task',
        'description': 'Missing several required fields',
        'priority': 0  # Invalid: must be >= 1
    }
    
    result = validator.validate_task(incomplete_task)
    
    print(f"Validation result: {'✓ Valid' if result.is_valid else '✗ Invalid'}")
    
    if result.errors:
        print(f"\nFound {len(result.errors)} errors:")
        for error in result.errors:
            print(f"  - {error}")
    
    print()


def example_6_implied_constraints():
    """Example 6: Detecting implied constraints and criteria"""
    print("=" * 60)
    print("Example 6: Implied Constraints Detection")
    print("=" * 60)
    
    validator = create_validator()
    
    task = {
        'task_name': 'Security implementation',
        'task_type': 'implementation',
        'description': 'Implement security features',
        'parameters': {},
        'status': 'PENDING',
        'priority': 1,
        'task_notes': '',
        'task_data': {
            'requirement_snippets': [],
            'requirement_type': 'security',
            'constraints': [],
            'success_criteria': [],
            'execution_complexity': 'complex',
            'project_id': 1,
            'document_ids': [1],
            'chunk_ids': [1],
            'dependencies': []
        }
    }
    
    # Chunk not marked but contains constraint language
    chunk = RAGChunk(
        chunk_id=1,
        chunk_text="""
        The API must not allow unauthorized access.
        All data must be encrypted at rest.
        User should be able to enable two-factor authentication.
        """,
        document_id=1,
        project_id=1,
        has_constraints=False,  # Not marked!
        has_success_criteria=False  # Not marked!
    )
    
    enriched_task, result = validator.validate_and_enrich_task(task, [chunk])
    
    print(f"Extracted {len(result.extracted_constraints)} constraints (implied)")
    print(f"Extracted {len(result.extracted_success_criteria)} success criteria (implied)")
    
    if result.warnings:
        print(f"\nWarnings ({len(result.warnings)}):")
        for warning in result.warnings:
            print(f"  ⚠ {warning}")
    
    print()


def example_7_bdd_style_criteria():
    """Example 7: BDD-style success criteria"""
    print("=" * 60)
    print("Example 7: BDD-Style Success Criteria")
    print("=" * 60)
    
    validator = create_validator()
    
    task = {
        'task_name': 'Shopping cart feature',
        'task_type': 'feature',
        'description': 'Implement shopping cart',
        'parameters': {},
        'status': 'PENDING',
        'priority': 2,
        'task_notes': '',
        'task_data': {
            'requirement_snippets': [],
            'requirement_type': 'functional',
            'constraints': [],
            'success_criteria': [],
            'execution_complexity': 'complex',
            'project_id': 1,
            'document_ids': [1],
            'chunk_ids': [1],
            'dependencies': []
        }
    }
    
    # BDD-style requirements
    chunk = RAGChunk(
        chunk_id=1,
        chunk_text="""
        Given a user with items in cart
        When they proceed to checkout
        Then they should see order summary
        And they should be able to complete purchase
        
        Given an empty cart
        When user tries to checkout
        Then system should display error message
        """,
        document_id=1,
        project_id=1,
        has_constraints=False,
        has_success_criteria=True
    )
    
    enriched_task, result = validator.validate_and_enrich_task(task, [chunk])
    
    print(f"Extracted {len(result.extracted_success_criteria)} BDD-style criteria:")
    for i, criterion in enumerate(result.extracted_success_criteria, 1):
        print(f"  {i}. {criterion}")
    
    print()


def example_8_real_world_scenario():
    """Example 8: Real-world task generation scenario"""
    print("=" * 60)
    print("Example 8: Real-World Scenario")
    print("=" * 60)
    
    validator = create_validator()
    
    # Simulating a real requirement document
    requirement_text = """
    User Authentication System Requirements
    
    The system must implement secure user authentication using industry-standard practices.
    
    Security Constraints:
    - Passwords must be hashed using bcrypt with minimum 10 rounds
    - Session tokens must expire after 30 minutes of inactivity
    - The system shall enforce HTTPS for all authentication endpoints
    - Maximum login attempts are limited to 5 per hour per IP
    - Multi-factor authentication is required for admin users
    
    Acceptance Criteria:
    - User must be able to register with email and password
    - User must be able to login with valid credentials
    - System should validate password strength (min 8 chars, 1 uppercase, 1 number)
    - Invalid login attempts must be logged
    - User should receive email confirmation after registration
    - Password reset functionality must be available
    
    When a user exceeds maximum login attempts, then their account should be temporarily locked.
    Given a valid session token, when user makes authenticated request, then system should grant access.
    """
    
    # Create task structure
    task = {
        'task_name': 'Implement User Authentication System',
        'task_type': 'feature',
        'description': 'Complete implementation of secure user authentication',
        'parameters': {
            'includes': ['registration', 'login', 'password_reset', 'mfa']
        },
        'status': 'PENDING',
        'priority': 1,
        'task_notes': 'Critical security feature - requires thorough testing',
        'task_data': {
            'requirement_snippets': [requirement_text],
            'requirement_type': 'functional',
            'constraints': [],
            'success_criteria': [],
            'execution_complexity': 'complex',
            'project_id': 1,
            'document_ids': [1],
            'chunk_ids': [1],
            'dependencies': ['Setup database schema', 'Configure email service']
        }
    }
    
    # Create RAG chunk from requirement
    chunk = RAGChunk(
        chunk_id=1,
        chunk_text=requirement_text,
        document_id=1,
        project_id=1,
        has_constraints=True,
        has_success_criteria=True,
        requirement_type='functional',
        priority_indicators='critical'
    )
    
    # Validate and enrich
    enriched_task, result = validator.validate_and_enrich_task(task, [chunk])
    
    print(f"Task: {enriched_task['task_name']}")
    print(f"Status: {'✓ Valid' if result.is_valid else '✗ Invalid'}")
    print(f"Priority: {enriched_task['priority']}")
    print(f"Complexity: {enriched_task['task_data']['execution_complexity']}")
    
    print(f"\nExtracted {len(result.extracted_constraints)} Constraints:")
    for i, constraint in enumerate(result.extracted_constraints[:5], 1):  # Show first 5
        print(f"  {i}. {constraint}")
    if len(result.extracted_constraints) > 5:
        print(f"  ... and {len(result.extracted_constraints) - 5} more")
    
    print(f"\nExtracted {len(result.extracted_success_criteria)} Success Criteria:")
    for i, criterion in enumerate(result.extracted_success_criteria[:5], 1):  # Show first 5
        print(f"  {i}. {criterion}")
    if len(result.extracted_success_criteria) > 5:
        print(f"  ... and {len(result.extracted_success_criteria) - 5} more")
    
    print(f"\nDependencies: {', '.join(enriched_task['task_data']['dependencies'])}")
    
    print()


def main():
    """Run all examples"""
    print("\n")
    print("╔" + "=" * 58 + "╗")
    print("║" + " " * 10 + "TASK SCHEMA VALIDATOR EXAMPLES" + " " * 17 + "║")
    print("╚" + "=" * 58 + "╝")
    print()
    
    example_1_basic_validation()
    example_2_constraint_extraction()
    example_3_success_criteria_extraction()
    example_4_multiple_chunks()
    example_5_validation_errors()
    example_6_implied_constraints()
    example_7_bdd_style_criteria()
    example_8_real_world_scenario()
    
    print("=" * 60)
    print("All examples completed!")
    print("=" * 60)


if __name__ == '__main__':
    main()

File: example-usage.js
/**
 * Example Usage: RAG-Aligned Task Generation Prompt Builder
 * 
 * This file demonstrates practical usage of the task prompt builder
 * in a real-world scenario.
 */

import {
  buildTaskGenerationPrompt,
  validateTaskList,
  extractRAGMetadata,
} from './taskPromptBuilder.js';

// ============================================================================
// Example 1: Basic Task Generation with Mock RAG Chunks
// ============================================================================

console.log('=== Example 1: Basic Task Generation ===\n');

const mockRAGChunks = [
  {
    document_id: 1,
    chunk_id: 5,
    requirement_type: 'functional',
    has_constraints: true,
    has_success_criteria: true,
    project_id: 42,
    document_name: 'API Requirements Document',
    similarity_score: 0.95,
    content: `The system must provide secure user authentication via JWT tokens. 
    The login endpoint must validate user credentials against the database and return 
    a signed JWT token upon successful authentication. The token must expire after 24 hours.`,
    extracted_constraints: [
      'Must use JWT tokens for authentication',
      'Must validate credentials against database',
      'Token must expire after 24 hours',
    ],
    extracted_success_criteria: [
      'User can successfully log in with valid credentials',
      'Invalid credentials return 401 error',
      'JWT token is returned on successful authentication',
      'Token expires after 24 hours',
    ],
  },
  {
    document_id: 1,
    chunk_id: 7,
    requirement_type: 'constraint',
    has_constraints: true,
    has_success_criteria: false,
    project_id: 42,
    document_name: 'API Requirements Document',
    similarity_score: 0.88,
    content: `All API endpoints must implement rate limiting to prevent abuse. 
    The system must enforce a maximum of 100 requests per minute per IP address. 
    Requests exceeding this limit must return a 429 Too Many Requests status code.`,
    extracted_constraints: [
      'Must implement rate limiting on all endpoints',
      'Maximum 100 requests per minute per IP address',
      'Must return 429 status code when limit exceeded',
    ],
  },
  {
    document_id: 1,
    chunk_id: 12,
    requirement_type: 'functional',
    has_constraints: false,
    has_success_criteria: true,
    project_id: 42,
    document_name: 'API Requirements Document',
    similarity_score: 0.85,
    content: `The system must provide a logout endpoint that invalidates the user's JWT token. 
    The endpoint must accept the token in the Authorization header and add it to a blacklist.`,
    extracted_success_criteria: [
      'User can successfully log out',
      'Token is invalidated after logout',
      'Invalidated token cannot be used for subsequent requests',
    ],
  },
  {
    document_id: 2,
    chunk_id: 3,
    requirement_type: 'acceptance_criteria',
    has_constraints: false,
    has_success_criteria: true,
    project_id: 42,
    document_name: 'Testing Requirements',
    similarity_score: 0.82,
    content: `The authentication system must be tested with at least 95% code coverage. 
    All edge cases must be covered including expired tokens, invalid tokens, and 
    concurrent login attempts. Load testing must verify the system can handle 
    1000 concurrent users.`,
    extracted_success_criteria: [
      'Achieve at least 95% code coverage for authentication module',
      'All edge cases are tested (expired, invalid, concurrent)',
      'System handles 1000 concurrent users in load tests',
    ],
  },
  {
    document_id: 3,
    chunk_id: 8,
    requirement_type: 'functional',
    has_constraints: true,
    has_success_criteria: true,
    project_id: 42,
    document_name: 'Database Schema',
    similarity_score: 0.79,
    content: `The user table must store hashed passwords using bcrypt with a cost factor of 12. 
    The table must include fields for email (unique), password_hash, created_at, and last_login. 
    Email addresses must be validated and stored in lowercase.`,
    extracted_constraints: [
      'Must use bcrypt with cost factor 12 for password hashing',
      'Email must be unique',
      'Email must be stored in lowercase',
    ],
    extracted_success_criteria: [
      'Passwords are securely hashed with bcrypt',
      'Email uniqueness is enforced at database level',
      'Email validation prevents invalid addresses',
    ],
  },
];

// Extract metadata for analysis
const metadata = extractRAGMetadata(mockRAGChunks);
console.log('RAG Context Analysis:');
console.log(`- Total chunks: ${metadata.totalChunks}`);
console.log(`- Source documents: ${metadata.uniqueDocuments}`);
console.log(`- Chunks with constraints: ${metadata.hasConstraints}`);
console.log(`- Chunks with success criteria: ${metadata.hasSuccessCriteria}`);
console.log('- Requirement types:');
Object.entries(metadata.requirementTypes).forEach(([type, count]) => {
  console.log(`  - ${type}: ${count}`);
});
console.log();

// Build the task generation prompt
const prompt = buildTaskGenerationPrompt({
  ragChunks: mockRAGChunks,
  taskQuery: `Generate a comprehensive, implementation-ready task list for building 
  the user authentication system. Include all necessary implementation tasks, 
  database setup, testing, and security considerations. Prioritize security-critical 
  tasks and ensure all constraints from the requirements are captured.`,
  projectId: 42,
  documentId: 1,
  agentId: 'cursor-agent-123',
  dbAgentId: 'db-agent-456',
  agentInstanceId: 'instance-789',
});

console.log('Generated Prompt Preview (first 1000 characters):');
console.log(prompt.substring(0, 1000));
console.log('...\n');
console.log(`Full prompt length: ${prompt.length} characters\n`);

// ============================================================================
// Example 2: Mock Task Generation and Validation
// ============================================================================

console.log('=== Example 2: Task Validation ===\n');

// Simulate tasks that might be generated by an LLM
const mockGeneratedTasks = [
  {
    task_name: 'Design and implement user authentication database schema',
    task_type: 'implementation',
    description: `Create the database schema for user authentication as specified in the requirements. 
    This includes creating a users table with fields for email (unique), password_hash, created_at, 
    and last_login. The schema must enforce email uniqueness at the database level and support 
    bcrypt password hashing with cost factor 12. Email addresses must be validated and stored in lowercase.`,
    parameters: {
      database: 'postgresql',
      table_name: 'users',
      fields: [
        { name: 'id', type: 'SERIAL PRIMARY KEY' },
        { name: 'email', type: 'VARCHAR(255) UNIQUE NOT NULL' },
        { name: 'password_hash', type: 'VARCHAR(255) NOT NULL' },
        { name: 'created_at', type: 'TIMESTAMP DEFAULT CURRENT_TIMESTAMP' },
        { name: 'last_login', type: 'TIMESTAMP' },
      ],
      indexes: ['email'],
    },
    status: 'PENDING',
    priority: 1,
    task_notes: 'Consider adding indexes for performance. Ensure migration is reversible.',
    task_data: {
      requirement_snippets: [
        'The user table must store hashed passwords using bcrypt with a cost factor of 12.',
        'The table must include fields for email (unique), password_hash, created_at, and last_login.',
        'Email addresses must be validated and stored in lowercase.',
      ],
      requirement_type: 'functional',
      constraints: [
        'Must use bcrypt with cost factor 12 for password hashing',
        'Email must be unique',
        'Email must be stored in lowercase',
      ],
      success_criteria: [
        'Passwords are securely hashed with bcrypt',
        'Email uniqueness is enforced at database level',
        'Email validation prevents invalid addresses',
      ],
      execution_complexity: 'medium',
      project_id: 42,
      document_ids: [3],
      chunk_ids: [8],
    },
    dependencies: [],
  },
  {
    task_name: 'Implement JWT-based login endpoint',
    task_type: 'implementation',
    description: `Implement the POST /api/auth/login endpoint that validates user credentials 
    and returns a signed JWT token. The endpoint must validate the email and password against 
    the database, hash the provided password using bcrypt, and compare it with the stored hash. 
    Upon successful authentication, generate a JWT token with 24-hour expiration and return it 
    to the client. Failed authentication attempts must return a 401 Unauthorized status.`,
    parameters: {
      endpoint: '/api/auth/login',
      method: 'POST',
      request_body: {
        email: 'string',
        password: 'string',
      },
      response: {
        success: { token: 'string', expires_at: 'timestamp' },
        error: { message: 'string', status: 401 },
      },
      token_expiration: '24h',
    },
    status: 'PENDING',
    priority: 1,
    task_notes: 'Implement rate limiting to prevent brute force attacks. Log failed attempts.',
    task_data: {
      requirement_snippets: [
        'The system must provide secure user authentication via JWT tokens.',
        'The login endpoint must validate user credentials against the database and return a signed JWT token upon successful authentication.',
        'The token must expire after 24 hours.',
      ],
      requirement_type: 'functional',
      constraints: [
        'Must use JWT tokens for authentication',
        'Must validate credentials against database',
        'Token must expire after 24 hours',
      ],
      success_criteria: [
        'User can successfully log in with valid credentials',
        'Invalid credentials return 401 error',
        'JWT token is returned on successful authentication',
        'Token expires after 24 hours',
      ],
      execution_complexity: 'high',
      project_id: 42,
      document_ids: [1],
      chunk_ids: [5],
    },
    dependencies: ['Design and implement user authentication database schema'],
  },
  {
    task_name: 'Implement rate limiting middleware',
    task_type: 'implementation',
    description: `Implement rate limiting middleware that enforces a maximum of 100 requests 
    per minute per IP address across all API endpoints. The middleware must track request counts 
    per IP address and return a 429 Too Many Requests status code when the limit is exceeded. 
    Include appropriate headers (X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset) 
    in responses to inform clients of their rate limit status.`,
    parameters: {
      max_requests: 100,
      time_window: '1 minute',
      tracking_key: 'ip_address',
      response_status: 429,
      headers: [
        'X-RateLimit-Limit',
        'X-RateLimit-Remaining',
        'X-RateLimit-Reset',
      ],
    },
    status: 'PENDING',
    priority: 2,
    task_notes: 'Consider using Redis for distributed rate limiting in production.',
    task_data: {
      requirement_snippets: [
        'All API endpoints must implement rate limiting to prevent abuse.',
        'The system must enforce a maximum of 100 requests per minute per IP address.',
        'Requests exceeding this limit must return a 429 Too Many Requests status code.',
      ],
      requirement_type: 'constraint',
      constraints: [
        'Must implement rate limiting on all endpoints',
        'Maximum 100 requests per minute per IP address',
        'Must return 429 status code when limit exceeded',
      ],
      success_criteria: [],
      execution_complexity: 'medium',
      project_id: 42,
      document_ids: [1],
      chunk_ids: [7],
    },
    dependencies: [],
  },
  {
    task_name: 'Implement logout endpoint with token blacklisting',
    task_type: 'implementation',
    description: `Implement the POST /api/auth/logout endpoint that invalidates the user's JWT token. 
    The endpoint must accept the token in the Authorization header, validate it, and add it to a 
    blacklist to prevent further use. The blacklist should be stored in a fast-access data store 
    (e.g., Redis) with automatic expiration matching the token's original expiration time.`,
    parameters: {
      endpoint: '/api/auth/logout',
      method: 'POST',
      headers: {
        Authorization: 'Bearer <token>',
      },
      blacklist_storage: 'redis',
      blacklist_ttl: 'token_expiration',
    },
    status: 'PENDING',
    priority: 2,
    task_notes: 'Ensure blacklist is checked on every authenticated request.',
    task_data: {
      requirement_snippets: [
        'The system must provide a logout endpoint that invalidates the user\'s JWT token.',
        'The endpoint must accept the token in the Authorization header and add it to a blacklist.',
      ],
      requirement_type: 'functional',
      constraints: [],
      success_criteria: [
        'User can successfully log out',
        'Token is invalidated after logout',
        'Invalidated token cannot be used for subsequent requests',
      ],
      execution_complexity: 'medium',
      project_id: 42,
      document_ids: [1],
      chunk_ids: [12],
    },
    dependencies: ['Implement JWT-based login endpoint'],
  },
  {
    task_name: 'Implement comprehensive authentication test suite',
    task_type: 'testing',
    description: `Create a comprehensive test suite for the authentication system that achieves 
    at least 95% code coverage. Tests must cover all edge cases including expired tokens, invalid 
    tokens, concurrent login attempts, rate limiting, and token blacklisting. Include unit tests 
    for individual components and integration tests for complete authentication flows. Implement 
    load tests to verify the system can handle 1000 concurrent users.`,
    parameters: {
      test_framework: 'jest',
      coverage_target: 95,
      test_types: ['unit', 'integration', 'load'],
      load_test_users: 1000,
      edge_cases: [
        'expired_tokens',
        'invalid_tokens',
        'concurrent_logins',
        'rate_limiting',
        'token_blacklisting',
      ],
    },
    status: 'PENDING',
    priority: 3,
    task_notes: 'Run load tests in staging environment before production deployment.',
    task_data: {
      requirement_snippets: [
        'The authentication system must be tested with at least 95% code coverage.',
        'All edge cases must be covered including expired tokens, invalid tokens, and concurrent login attempts.',
        'Load testing must verify the system can handle 1000 concurrent users.',
      ],
      requirement_type: 'acceptance_criteria',
      constraints: [],
      success_criteria: [
        'Achieve at least 95% code coverage for authentication module',
        'All edge cases are tested (expired, invalid, concurrent)',
        'System handles 1000 concurrent users in load tests',
      ],
      execution_complexity: 'high',
      project_id: 42,
      document_ids: [2],
      chunk_ids: [3],
    },
    dependencies: [
      'Implement JWT-based login endpoint',
      'Implement logout endpoint with token blacklisting',
      'Implement rate limiting middleware',
    ],
  },
];

// Validate the generated tasks
console.log('Validating generated tasks...\n');
const validation = validateTaskList(mockGeneratedTasks);

if (validation.valid) {
  console.log('✓ All tasks passed validation!\n');
  console.log(`Generated ${mockGeneratedTasks.length} valid tasks:\n`);
  mockGeneratedTasks.forEach((task, index) => {
    console.log(`${index + 1}. ${task.task_name}`);
    console.log(`   Type: ${task.task_type}`);
    console.log(`   Priority: ${task.priority}`);
    console.log(`   Complexity: ${task.task_data.execution_complexity}`);
    console.log(`   Source: ${task.task_data.document_ids.length} document(s), ${task.task_data.chunk_ids.length} chunk(s)`);
    console.log(`   Dependencies: ${task.dependencies.length > 0 ? task.dependencies.join(', ') : 'None'}`);
    console.log();
  });
} else {
  console.log('✗ Validation failed!\n');
  validation.taskErrors.forEach(error => {
    console.log(`Task: ${error.task_name}`);
    console.log(`Errors:`);
    error.errors.forEach(err => console.log(`  - ${err}`));
    console.log();
  });
}

// ============================================================================
// Example 3: Traceability Analysis
// ============================================================================

console.log('=== Example 3: Requirement Traceability Analysis ===\n');

// Analyze traceability of generated tasks
console.log('Traceability Matrix:\n');

const documentCoverage = new Map();
const chunkCoverage = new Map();

mockGeneratedTasks.forEach(task => {
  task.task_data.document_ids.forEach(docId => {
    if (!documentCoverage.has(docId)) {
      documentCoverage.set(docId, []);
    }
    documentCoverage.get(docId).push(task.task_name);
  });

  task.task_data.chunk_ids.forEach(chunkId => {
    if (!chunkCoverage.has(chunkId)) {
      chunkCoverage.set(chunkId, []);
    }
    chunkCoverage.get(chunkId).push(task.task_name);
  });
});

console.log('Document Coverage:');
documentCoverage.forEach((tasks, docId) => {
  const doc = mockRAGChunks.find(c => c.document_id === docId);
  console.log(`\nDocument ${docId}: ${doc?.document_name || 'Unknown'}`);
  console.log(`Tasks implementing requirements from this document:`);
  tasks.forEach(taskName => console.log(`  - ${taskName}`));
});

console.log('\n\nChunk Coverage:');
chunkCoverage.forEach((tasks, chunkId) => {
  const chunk = mockRAGChunks.find(c => c.chunk_id === chunkId);
  console.log(`\nChunk ${chunkId} (${chunk?.requirement_type || 'unknown'}):`);
  console.log(`Content preview: ${chunk?.content.substring(0, 80)}...`);
  console.log(`Tasks implementing this requirement:`);
  tasks.forEach(taskName => console.log(`  - ${taskName}`));
});

console.log('\n\n=== Examples Complete ===');
